File: kinotic-core/src/main/java/org/kinotic/core/api/event/DefaultCRI.java


package org.kinotic.core.api.event;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Locale;

/**
 *
 * Created by Navid Mitchell on 5/1/20
 */
class DefaultCRI implements CRI {

    private static final char ZONE_DELIMITER = '~';

    private final URI uri;

    public DefaultCRI(String scheme, String scope, String resourceName, String path, String version) {
        // '@' delimits the scope in the composed authority, so neither part may carry one
        if (scope != null && scope.indexOf('@') >= 0) {
            throw new IllegalArgumentException("The scope must not contain '@' but was '" + scope + "'");
        }
        if (resourceName != null && resourceName.indexOf('@') >= 0) {
            throw new IllegalArgumentException("The resourceName must not contain '@' but was '" + resourceName + "'");
        }
        // Compose the authority as scope@resourceName and pass it to the authority-form URI
        // constructor, which stores it verbatim. The scheme and authority are the CRI's identity,
        // so they are lowercased; the path and version are payload details and keep their case.
        String authority = resourceName != null
                ? ((scope != null ? scope + "@" : "") + resourceName).toLowerCase(Locale.ROOT)
                : null;
        try {
            uri = new URI(scheme.toLowerCase(Locale.ROOT), authority, path, null, version);
        } catch (URISyntaxException x) {
            throw new IllegalArgumentException(x.getMessage(), x);
        }
        validateIdentity();
    }

    /**
     * Will create a {@link CRI} from a raw string
     *
     * @param rawCRI the raw string to create from an {@link CRI}
     */
    public DefaultCRI(String rawCRI) {
        uri = URI.create(lowercaseIdentity(rawCRI));
        validateIdentity();
    }

    // '~' delimits the zone and '@' delimits the scope, so any other occurrence of either could
    // only confuse parsing — a delimiter inside a part is rejected outright
    private void validateIdentity() {
        String scope = scope();
        if (scope != null && scope.indexOf(ZONE_DELIMITER) >= 0) {
            throw new IllegalArgumentException("The scope must not contain '~' but was '" + scope + "'");
        }
        String resourceName = resourceName();
        if (resourceName != null && resourceName.indexOf(ZONE_DELIMITER) >= 0) {
            throw new IllegalArgumentException("The resourceName must not contain '~' but was '" + resourceName + "'");
        }
        String authority = uri.getRawAuthority();
        if (authority != null && authority.indexOf('@') != authority.lastIndexOf('@')) {
            throw new IllegalArgumentException("The authority must contain at most one '@' but was '" + authority + "'");
        }
    }

    // Lowercases the scheme and authority of a raw CRI, leaving the path and everything after it
    // untouched. The authority ends at the first '/', '?', or '#' after "://".
    private static String lowercaseIdentity(String rawCRI) {
        String ret = rawCRI;
        int schemeEnd = rawCRI.indexOf("://");
        if (schemeEnd >= 0) {
            int authorityEnd = rawCRI.length();
            for (int i = schemeEnd + 3; i < rawCRI.length(); i++) {
                char c = rawCRI.charAt(i);
                if (c == '/' || c == '?' || c == '#') {
                    authorityEnd = i;
                    break;
                }
            }
            ret = rawCRI.substring(0, authorityEnd).toLowerCase(Locale.ROOT) + rawCRI.substring(authorityEnd);
        }
        return ret;
    }

    @Override
    public String scheme() {
        return uri.getScheme();
    }

    @Override
    public String scope() {
        // The raw authority is scope@[zone~]resourceName; scope is the part before '@', null when absent.
        String authority = uri.getRawAuthority();
        if (authority == null) {
            return null;
        }
        int at = authority.indexOf('@');
        return at >= 0 ? authority.substring(0, at) : null;
    }

    @Override
    public boolean hasScope() {
        return scope() != null;
    }

    @Override
    public String zone() {
        String hostPart = hostPart();
        if (hostPart == null) {
            return null;
        }
        int delimiter = hostPart.indexOf(ZONE_DELIMITER);
        return delimiter >= 0 ? hostPart.substring(0, delimiter) : null;
    }

    @Override
    public boolean hasZone() {
        return zone() != null;
    }

    @Override
    public String resourceName() {
        String hostPart = hostPart();
        if (hostPart == null) {
            return null;
        }
        int delimiter = hostPart.indexOf(ZONE_DELIMITER);
        return delimiter >= 0 ? hostPart.substring(delimiter + 1) : hostPart;
    }

    // The authority after any scope@, holding [zone~]resourceName
    private String hostPart() {
        String authority = uri.getRawAuthority();
        if (authority == null) {
            return null;
        }
        int at = authority.indexOf('@');
        return at >= 0 ? authority.substring(at + 1) : authority;
    }

    @Override
    public String version() {
        return uri.getRawFragment();
    }

    @Override
    public boolean hasVersion() {
        return uri.getRawFragment() != null;
    }

    @Override
    public String path() {
        return uri.getRawPath();
    }

    @Override
    public boolean hasPath() {
        return uri.getRawPath() != null;
    }

    @Override
    public String baseResource() {
        StringBuilder sb = new StringBuilder(scheme());
        sb.append("://");
        if(hasScope()){
            sb.append(scope());
            sb.append("@");
        }
        if(hasZone()){
            sb.append(zone());
            sb.append(ZONE_DELIMITER);
        }
        sb.append(resourceName());

        return sb.toString();
    }

    @Override
    public String raw() {
        return uri.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;

        if (o == null || getClass() != o.getClass()) return false;

        DefaultCRI that = (DefaultCRI) o;

        return new EqualsBuilder()
                .append(uri.toString(), that.toString())
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .append(uri.toString())
                .toHashCode();
    }

    @Override
    public String toString() {
        return uri.toString();
    }
}

File: kinotic-js/workspace/packages/core/src/api/event/DefaultCRI.ts
import type {CRI} from "./CRI"

const ZONE_DELIMITER = '~'

/**
 * Default implementation of the `CRI` interface.
 *
 * @author Navid Mitchell
 * @since 3/25/25
 */
export class DefaultCRI implements CRI {
    private readonly _scheme: string
    private readonly _scope: string | null
    private readonly _zone: string | null
    private readonly _resourceName: string
    private readonly _path: string | null
    private readonly _version: string | null
    private readonly _raw: string

    constructor(rawCRI: string)
    constructor(scheme: string, scope: string | null, resourceName: string, path: string | null, version: string | null)
    constructor(...args: any[]) {
        // The scheme, scope, zone, and resourceName are the CRI's identity, so they are
        // lowercased; the path and version are payload details and keep their case.
        if (args.length === 1) {
            const rawURC = args[0]
            if (typeof rawURC !== "string") {
                throw new Error("Raw URI must be a string")
            }
            const parsed = DefaultCRI.parseRaw(rawURC)
            this._scheme = parsed.scheme.toLowerCase()
            this._scope = parsed.scope ? parsed.scope.toLowerCase() : null
            const [zone, resourceName] = DefaultCRI.splitZone(parsed.resourceName.toLowerCase())
            this._zone = zone
            this._resourceName = resourceName
            this._path = parsed.path
            this._version = parsed.version
        } else if (args.length === 5) {
            const [scheme, scope, resourceName, path, version] = args
            this._scheme = scheme.toLowerCase()
            this._scope = scope ? scope.toLowerCase() : null
            const [zone, name] = DefaultCRI.splitZone((resourceName as string).toLowerCase())
            this._zone = zone
            this._resourceName = name
            this._path = path
            this._version = version
        } else {
            throw new Error("Invalid constructor arguments for DefaultCRI")
        }
        this._raw = DefaultCRI.buildRaw(this._scheme, this._scope, this._zone, this._resourceName, this._path, this._version)

        if (!this._scheme || !this._resourceName) {
            throw new Error(`Invalid CRI: scheme and resourceName are required. Got: ${this._raw}`)
        }
        // '~' delimits the zone from the resourceName, so any other occurrence could only confuse
        // zone parsing — a scope carrying one, or a second '~' in the host part, is rejected outright
        if (this._scope !== null && this._scope.includes(ZONE_DELIMITER)) {
            throw new Error(`The scope must not contain '~' but was '${this._scope}'`)
        }
        if (this._resourceName.includes(ZONE_DELIMITER)) {
            throw new Error(`The resourceName must not contain '~' but was '${this._resourceName}'`)
        }
    }

    public scheme(): string {
        return this._scheme
    }

    public scope(): string | null {
        return this._scope
    }

    public hasScope(): boolean {
        return this._scope !== null
    }

    public zone(): string | null {
        return this._zone
    }

    public hasZone(): boolean {
        return this._zone !== null
    }

    public resourceName(): string {
        return this._resourceName
    }

    public version(): string | null {
        return this._version
    }

    public hasVersion(): boolean {
        return this._version !== null
    }

    public path(): string | null {
        return this._path
    }

    public hasPath(): boolean {
        return this._path !== null
    }

    public baseResource(): string {
        let result = `${this._scheme}://`
        if (this.hasScope()) {
            result += `${this._scope}@`
        }
        if (this.hasZone()) {
            result += `${this._zone}${ZONE_DELIMITER}`
        }
        result += this._resourceName
        return result
    }

    public raw(): string {
        return this._raw
    }

    public equals(other: any): boolean {
        if (this === other) return true
        if (!(other instanceof DefaultCRI)) return false
        return this._raw === other.raw()
    }

    public hashCode(): number {
        let hash = 17
        hash = hash * 37 + this._raw.split("").reduce((a, c) => a + c.charCodeAt(0), 0)
        return hash
    }

    public toString(): string {
        return this._raw
    }

    // Splits [zone~]resourceName into its zone (or null) and resourceName
    private static splitZone(hostPart: string): [string | null, string] {
        const delimiter = hostPart.indexOf(ZONE_DELIMITER)
        return delimiter >= 0
            ? [hostPart.substring(0, delimiter), hostPart.substring(delimiter + 1)]
            : [null, hostPart]
    }

    // Helper to parse a raw CRI string
    private static parseRaw(raw: string): {
        scheme: string
        scope: string | null
        resourceName: string
        path: string | null
        version: string | null
    } {
        const regex = /^([^:]+):\/\/(?:([^@]+)@)?([^\/#]+)(\/[^#]*)?(?:#(.+))?$/;
        const match = raw.match(regex)
        if (!match) {
            throw new Error(`Invalid CRI format: ${raw}`)
        }

        const [, scheme, scope, resourceName, path, version] = match
        return {
            scheme: scheme as string,
            scope: scope || null,
            resourceName: resourceName as string,
            path: path ? path.substring(1) : null,
            version: version || null,
        }
    }

    // Helper to build a raw CRI string
    private static buildRaw(
        scheme: string,
        scope: string | null,
        zone: string | null,
        resourceName: string,
        path: string | null,
        version: string | null
    ): string {
        let result = `${scheme}://`
        if (scope) {
            result += `${scope}@`
        }
        if (zone) {
            result += `${zone}${ZONE_DELIMITER}`
        }
        result += resourceName
        if (path) {
            result += `/${path}`
        }
        if (version) {
            result += `#${version}`
        }
        return result
    }
}
